Betaflight firmware support

nikhil-sethi opened the pull request:
<!-- Thank you for submitting a pull request! -->
<!-- ⚠️⚠️ Do Not Delete This! pull_request_template ⚠️⚠️ -->
<!-- Please read our contribution guidelines: https://microsoft.github.io/AirSim/CONTRIBUTING/ -->

Fixes: 
NA

## About
<!-- Describe what your PR is about. -->
This PR adds support for betaflight as a new firmware for drone flight. Reference issue: #4351 
The implementation is similar to that of arducopter with some basic changes that I wanted to discuss.

## How Has This Been Tested?
<!-- Please, describe how you have tested your changes to help us incorporate them. -->
- Tested with betaflight 4.3.0 at commit 3267f041
- Currently only tested on Linux (Ubuntu 20.04). I understand that windows is recommended but my dual boot does not have AirSim + Unreal setup as of yet. Let me know if I should set it up or if someone else can test it! Also, I don't have a mac machine.
- RC input tested with a very basic [custom program](https://github.com/nikhil-sethi/joystick) that outputs normalised joystick values to a UDP socket that betaflight listens to.

Known issues and doubts:
1. The stick responses are good enough for all axes after tuning PID values but the throttle seems sluggish for flying or testing. (takeoff at ~1700; almost full when doing maneuvers). Is this a basic inertia/mass issue? Shouldn't it be decoupled from raw PWM values. I haven't investigated this yet.
2. The arducopter implementation uses json to get the fdm packet here and there. #2075 says this is for easy addition and removal of sensors. However, for better compatibility, I found it easier and minimal to implement an fdm packet struct equivalent to the SITL [target](https://github.com/betaflight/betaflight/blob/7b4415f062656218427ee160b776ac5488eca5f1/src/main/target/SITL/target.h#L238) in betaflight. Are there some performance issues with this implementation that would be better with json? Please let me know if I am missing something.

## Screenshots (if appropriate):
- Testing video: https://youtu.be/sVEfow3y6rU

## TODO
- [ ] Fix/understand throttle issue
- [x] Fix/Understand fdm data structure
- [x] Format and cleanup according to repo style

nikhil-sethi committed:
Added betaflight firmware
- tested with betaflight 4.3.0 at commit 3267f041
- minor issues with throttle
- formatting required

nikhil-sethi committed:
Formatting + small fixes
- timestamp fix: changed to seconds

rajat2004 commented on AirLib/include/vehicles/multirotor/firmwares/betaflight/BetaflightApi.hpp:
```diff
@@ -0,0 +1,413 @@
+            }
+
+            Utils::log(Utils::stringf("Using UDP port %d, local IP %s, remote IP %s for sending sensor data", port_, connection_info_.local_host_ip.c_str(), ip_.c_str()), Utils::kLogLevelInfo);
+            Utils::log(Utils::stringf("Using UDP port %d for receiving rotor power", connection_info_.control_port_local, connection_info_.local_host_ip.c_str(), ip_.c_str()), Utils::kLogLevelInfo);
```
The log string doesn't have the required `%s` here to print the vars

nikhil-sethi replied:
Yes. Will fix this.

rajat2004 commented on AirLib/include/vehicles/multirotor/firmwares/betaflight/BetaflightApi.hpp:
```diff
@@ -0,0 +1,413 @@
+            fdmPacket pkt;
+
+            // Time
+            pkt.timestamp = ClockFactory::get()->nowNanos() / 1000;
```
This seems wrong. First, the fdm definition mentions seconds, but this will give value in microseconds.
Also, the type is `double`, but the RHS `nowNanos()` returns uint64, so dividing by 10^3 will lose ns info (probably not needed), and then convert to double. Ensure the required precision is calculated correctly.

nikhil-sethi replied:
1. I agree that it's wrong and I knew it would get pointed out. But for some reason, the right implementation with division by 1E9 didn't seem to be working with betaflight. 
Upon further debugging I recently found this was a timeout issue in [this](https://github.com/betaflight/betaflight/blob/7b4415f062656218427ee160b776ac5488eca5f1/src/main/target/SITL/target.c#L94) line and the timeout should be increased to 600ms to calculate the correct simRate. There should be a general method to fix this on the betaflight side. Will work on it in the future.
Also betaflight needs relative time from start of simulation which I have changed.

2. I understand what you're saying. I could cast the time to a double before division like so:
```
pkt.timestamp = static_cast<double>(ClockFactory::get()->nowNanos()) / 1E9;
```
But this gives me the same result without casting and works the same with betaflight. This timestamp is used to find the simulation delta time which is exactly 0.003 sec. Am I understanding correctly that the lost ns info doesn't matter (when dividing by 1e9) because of this reason? I'll still add the casting to make it clearer though.

rajat2004 commented on AirLib/include/vehicles/multirotor/firmwares/betaflight/BetaflightApi.hpp:
```diff
@@ -0,0 +1,413 @@
+                // Position
+
+                // Velocity
+            if (udp_socket_ != nullptr) {
+                udp_socket_->sendto(&pkt, sizeof(pkt), ip_, port_);
+            }
```
This check could be done in the start itself, if `nullptr` then just return early

nikhil-sethi replied:
Agreed. Will add the check for sensors also at the top

rajat2004 commented on AirLib/include/vehicles/multirotor/firmwares/betaflight/BetaflightApi.hpp:
```diff
@@ -0,0 +1,413 @@
+            }
+
+            normalizeRotorControls();
+            Utils::log(Utils::stringf("pwm received: [ %f, %f, %f, %f ]", rotor_controls_[0], rotor_controls_[1], rotor_controls_[2], rotor_controls_[3]), Utils::kLogLevelInfo);
```
Should be removed once things are final, otherwise too much logging

nikhil-sethi replied:
Yes definitely. Will fix

rajat2004 commented on AirLib/include/vehicles/multirotor/firmwares/betaflight/BetaflightApi.hpp:
```diff
@@ -0,0 +1,413 @@
+        double imu_orientation_quat[4];     // w, x, y, z
+        double velocity_xyz[3];             // m/s, earth frame
+        double position_xyz[3];             // meters, NED from origin
+    } ;
```
Do confirm whether you want `__attribute__((__packed__))` to avoid padding issues when reading raw bytes directly into struct

nikhil-sethi replied:
I'm not familiar with this attribute but I understand what it's doing. I'll read up + experiment with it before confirming. Will get back to you for sure.

rajat2004 commented on AirLib/include/vehicles/multirotor/firmwares/betaflight/BetaflightApi.hpp:
```diff
@@ -0,0 +1,413 @@
+    static const int kBetaflightRotorControlCount = 4;
+    
+    struct servoPacket {  // equivalent of servo_packet in betaflight SITL
+        u_int16_t pwm[kBetaflightRotorControlCount]; //
```
What is this `u_int16_t`? Haven't seen this before...

nikhil-sethi replied:
(today I learned) Apparently this is (one of) the typedefs in sys/types.h used for integers before C99 came up with the standards for uintXX_t. I must've added an underscore by mistake and let intellisense worry about the rest. Cursing myself anyways..

nikhil-sethi replied:
Will correct it to the accepted standard.

rajat2004 reviewed (commented):
Just a brief look, would be great if you could also do the clang-formatting, setting up in IDE should be simple. The maintainers will need to grant access to run the Github workflows which would do some basic compilation and formatting checks

rajat2004 commented:
Re JSON vs FDM packet, the benefit with JSON (or any kind of serialized format) is that the sender and receiver can add more types of data, extract info that is required and ignore the rest (e.g. if something is not supported, backwards compatibility). For e.g. with Lidar & RC control support, it would have worked even if either side didn't support it, it would just be ignored. For the fixed struct, variable data like Lidar won't work very nicely since the receiver would need to know exactly where the data starts and ends. New fields can only be added at the end, any change anywhere else in the struct would cause incorrect data being read if older struct is being used on other side. Again one of the reasons with ArudPilot is also moving more towards using JSON as a generic and flexible way to implement interfaces for multiple simulators

nikhil-sethi commented:
> Re JSON vs FDM packet, the benefit with JSON (or any kind of serialized format) is that the sender and receiver can add more types of data, extract info that is required and ignore the rest (e.g. if something is not supported, backwards compatibility). For e.g. with Lidar & RC control support, it would have worked even if either side didn't support it, it would just be ignored. For the fixed struct, variable data like Lidar won't work very nicely since the receiver would need to know exactly where the data starts and ends. New fields can only be added at the end, any change anywhere else in the struct would cause incorrect data being read if older struct is being used on other side.
> 
> FDM struct would be faster in performance than serializing to string and later parsing it, it'll most likely not be the bottleneck for the simulation, there are other much more obvious issues like AirSim internal physics and UE4 rendering, firmware side handling, and then network communication. Another thing that was very useful with the JSON interface was the ease of debugging, the packet can be easily read manually in Wireshark, to spot any obvious errors.
> 
> ArduPilot is also moving towards using JSON for interfacing with multiple simulators, e.g. [this Discuss post](https://discuss.ardupilot.org/t/sitl-protocol-for-custom-simulators/60365/5)

Thanks for explaining this in detail. It makes sense now why JSON was used. I'll progress with FDM for current testing purposes and in the meantime discuss at the betaflight PR if I can add the JSON unpacking routine in their target.

nikhil-sethi committed:
Timestamp fixes + clang formatting
- added casting for timestamp + changed to relative time
- moved nullptr check to top  in updateState
- some logging fixes

zhouqingf commented:
Hello, in the process of replicating airsim and bateflight, I couldn't implement communication between airsim and silt. I tried to configure setting.json, but once tested, UE waited for a communication connection and the program wouldn't work. I'd like to know how your setting.jsons are configured.
